overlord/snapstate/autorefresh_gating.go

```diff
@@ -0,0 +1,223 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package snapstate
+
+import (
+	"fmt"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/mount"
+	"github.com/snapcore/snapd/overlord/ifacestate/ifacerepo"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/release"
+	"github.com/snapcore/snapd/snap"
+)
+
+var gateAutoRefreshHookName = "gate-auto-refresh"
+
+type affectedSnapInfo struct {
+	Restart        bool
+	Base           bool
+	AffectingSnaps map[string]bool
+}
+
+func affectedByRefresh(st *state.State, updates []*snap.Info) (map[string]*affectedSnapInfo, error) {
+	all, err := All(st)
+	if err != nil {
+		return nil, err
+	}
+
+	var bootBase string
+	if !release.OnClassic {
+		deviceCtx, err := DeviceCtx(st, nil, nil)
+		if err != nil {
+			return nil, fmt.Errorf("cannot get device context: %v", err)
+		}
+		bootBaseInfo, err := BootBaseInfo(st, deviceCtx)
+		if err != nil {
+			return nil, fmt.Errorf("cannot get boot base info: %v", err)
+		}
+		bootBase = bootBaseInfo.InstanceName()
+	}
+
+	byBase := make(map[string][]string)
+	for name, snapSt := range all {
+		if !snapSt.Active {
+			delete(all, name)
+			continue
+		}
+		inf, err := snapSt.CurrentInfo()
+		if err != nil {
+			return nil, err
+		}
+		// optimize: do not consider snaps that don't have gate-auto-refresh hook.
+		if inf.Hooks[gateAutoRefreshHookName] == nil {
+			delete(all, name)
+			continue
+		}
+
+		base := inf.Base
+		if base == "none" {
+			continue
+		}
+		if inf.Base == "" {
+			base = "core"
+		}
+		byBase[base] = append(byBase[base], inf.InstanceName())
+	}
+
+	affected := make(map[string]*affectedSnapInfo)
+
+	addAffected := func(snapName, affectedBy string, restart bool, base bool) {
+		affectedInfo := affected[snapName]
+		if affectedInfo == nil {
+			affectedInfo = &affectedSnapInfo{
+				AffectingSnaps: map[string]bool{},
+			}
+			affected[snapName] = affectedInfo
+		}
+		if restart {
+			affectedInfo.Restart = restart
+		}
+		if base {
+			affectedInfo.Base = base
+		}
+		affectedInfo.AffectingSnaps[affectedBy] = true
+	}
+
+	for _, up := range updates {
+		// on core system, affected by update of boot base
+		if bootBase != "" && up.InstanceName() == bootBase {
+			for _, snapSt := range all {
+				addAffected(snapSt.InstanceName(), up.InstanceName(), true, false)
```

Comment:
(nitpick) bool parameters like this are not my favorite because I always need to look at the signature to remember if true means reboot or base here. But given that this is local api only and it's all close together it's fine.